### PR TITLE
Added an optional parameter dashscope_api_key

### LINKED ISSUE #14739: [Question]: Why do some llama-index integrations not allow the 'api key' and the 'base url' to be passed in as an initialization parameter?
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

Some integrations only allow you to set an api key and a base url as environment variables but this limits the utility of a class since the developer can only initialize one such client. A suggestion would be to have the integrations accept these attributes as optional parameters and then default to values set in the environment if they are not provided. 

In some instances, the consumer is forced to wrap these integrations in classes which accept these parameters and this is fine but these types of integrations which do not accept api keys are increasing. Most of the llama-index-readers-google abstractions as well as the dashcope node parser are a few examples that come to mind.

### PR DESCRIPTION
**Pull request by ShashankVSG**
# Description

Added an optional parameter dashscope_api_key to initialize the DashscopeNodeParser class.

Fixes #14739 (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [ ] Yes
- [x] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [ ] Yes
- [x] No

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py [hunk 1]
```diff
@@ -36,6 +36,9 @@ class DashScopeJsonNodeParser(BaseElementNodeParser):
         description="language of tokenizor, accept cn, en, any. Notice that <any> mode will be slow.",
     )
 
+    def __init__(self, dashscope_api_key: Optional[str] = None):
+        self.dashscope_api_key = dashscope_api_key or os.environ.get("DASHSCOPE_API_KEY")
```
Comment (nerdai) [inline llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py:40]: I think we should either set this as a `Field` like the others or set as a `PrivateAttr` and rename to `self._dashscope_api_key`.

Something like this: https://github.com/run-llama/llama_index/blob/f1965dfad3fbf9ff1e97e1948a4ed8604c9c266b/llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py#L100

**Review by nerdai — COMMENTED** [review]
Thanks @ShashankVSG  -- left a suggested edit in my review.

File: llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py (hunk 1 above)
Comment (ShashankVSG) [reply llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py]: Modified the dashscope_api_key to PrivateAttr()

File: llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py [hunk 2]
```diff
@@ -35,6 +35,10 @@ class DashScopeJsonNodeParser(BaseElementNodeParser):
         default="cn",
         description="language of tokenizor, accept cn, en, any. Notice that <any> mode will be slow.",
     )
+    _dashscope_api_key: str = PrivateAttr()
+
+    def __init__(self, dashscope_api_key: Optional[str] = None):
+        self._dashscope_api_key = dashscope_api_key or os.environ.get("DASHSCOPE_API_KEY")
```
Comment (nerdai) [inline llama-index-integrations/node_parser/llama-index-node-parser-relational-dashscope/llama_index/node_parser/dashscope/base.py:41]: I'm almost certain here that Pydantic will complain if we don't call `super().__init__(...)` with the correct params.

You can follow the example I previous linked again for this: https://github.com/run-llama/llama_index/blob/f1965dfad3fbf9ff1e97e1948a4ed8604c9c266b/llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py#L134

**Pull request closed without merging**